File: .config/nvim/lua/plugins/coc.lua
return {
  "neoclide/coc.nvim",

  enabled = vim.g.plugin_use_coc,

  branch = "release",

  event = "VeryLazy",

  init = function()
    vim.g.coc_global_extensions = {
      "coc-deno",
      "coc-tsserver",
      "coc-diagnostic",
      "coc-explorer",
      "coc-json",
      "coc-lists",
      "coc-marketplace",
      "coc-mocword",
      "coc-nav",
      "coc-powershell",
      "coc-lua",
      "coc-stylua",
      "coc-rust-analyzer",
      "coc-spell-checker",
      "coc-translator",
      "coc-highlight",
      "coc-tsdetect",
      "coc-vimlsp",
    }

    local keyset = vim.keymap.set
    -- Autocomplete
    function _G.check_back_space()
      local col = vim.fn.col(".") - 1
      return col == 0 or vim.fn.getline("."):sub(col, col):match("%s") ~= nil
    end

    -- Use Tab for trigger completion with characters ahead and navigate
    -- NOTE: There's always a completion item selected by default, you may want to enable
    -- no select by setting `"suggest.noselect": true` in your configuration file
    -- NOTE: Use command ':verbose imap <tab>' to make sure Tab is not mapped by
    -- other plugins before putting this into your config
    local opts = { silent = true, noremap = true, expr = true, replace_keycodes = false }
    keyset(
      "i",
      "<c-n>",
      'coc#pum#visible() ? coc#pum#next(1) : v:lua.check_back_space() ? "<c-n>" : coc#refresh()',
      opts
    )
    keyset("i", "<c-p>", [[coc#pum#visible() ? coc#pum#prev(1) : "\<c-p>"]], opts)

    -- Make <CR> to accept selected completion item or notify coc.nvim to format
    -- <C-g>u breaks current undo, please make your own choice
    -- keyset("i", "<cr>", [[coc#pum#visible() ? coc#pum#confirm() : "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"]], opts)

    -- Use <c-j> to trigger snippets
    keyset("i", "<c-j>", "<Plug>(coc-snippets-expand-jump)")
    -- Use <c-space> to trigger completion
    keyset("i", "<c-space>", "coc#refresh()", { silent = true, expr = true })

    -- Use `[d` and `]d` to navigate diagnostics
    -- Use `:CocDiagnostics` to get all diagnostics of current buffer in location list
    keyset("n", "[d", "<Plug>(coc-diagnostic-prev)", { silent = true })
    keyset("n", "]d", "<Plug>(coc-diagnostic-next)", { silent = true })

    -- GoTo code navigation
    keyset("n", "gd", "<Plug>(coc-definition)", { silent = true })
    keyset("n", "gy", "<Plug>(coc-type-definition)", { silent = true })
    keyset("n", "gi", "<Plug>(coc-implementation)", { silent = true })
    keyset("n", "gr", "<Plug>(coc-references)", { silent = true })

    -- Use K to show documentation in preview window
    function _G.show_docs()
      local cw = vim.fn.expand("<cword>")
      if vim.fn.index({ "vim", "help" }, vim.bo.filetype) >= 0 then
        vim.api.nvim_command("h " .. cw)
      elseif vim.api.nvim_eval("coc#rpc#ready()") then
        vim.fn.CocActionAsync("doHover")
      else
        vim.api.nvim_command("!" .. vim.o.keywordprg .. " " .. cw)
      end
    end

    keyset("n", "K", "<CMD>lua _G.show_docs()<CR>", { silent = true })

    -- Highlight the symbol and its references on a CursorHold event(cursor is idle)
    vim.api.nvim_create_augroup("CocGroup", {})
    vim.api.nvim_create_autocmd("CursorHold", {
      group = "CocGroup",
      command = "silent call CocActionAsync('highlight')",
      desc = "Highlight symbol under cursor on CursorHold",
    })

    -- Symbol renaming
    keyset("n", "<leader>rn", "<Plug>(coc-rename)", { silent = true })

    -- Formatting selected code
    keyset("x", "<leader>F", "<Plug>(coc-format-selected)", { silent = true })
    keyset("n", "<leader>F", "<Plug>(coc-format-selected)", { silent = true })

    -- Setup formatexpr specified filetype(s)
    vim.api.nvim_create_autocmd("FileType", {
      group = "CocGroup",
      pattern = "typescript,json",
      command = "setl formatexpr=CocAction('formatSelected')",
      desc = "Setup formatexpr specified filetype(s).",
    })

    -- Update signature help on jump placeholder
    vim.api.nvim_create_autocmd("User", {
      group = "CocGroup",
      pattern = "CocJumpPlaceholder",
      command = "call CocActionAsync('showSignatureHelp')",
      desc = "Update signature help on jump placeholder",
    })

    -- Apply codeAction to the selected region
    -- Example: `<leader>aap` for current paragraph
    local opts = { silent = true, nowait = true }
    keyset("x", "<leader>a", "<Plug>(coc-codeaction-selected)", opts)
    keyset("n", "<leader>a", "<Plug>(coc-codeaction-selected)", opts)

    -- Remap keys for apply code actions at the cursor position.
    keyset("n", "<leader>ac", "<Plug>(coc-codeaction-cursor)", opts)
    -- Remap keys for apply code actions affect whole buffer.
    keyset("n", "<leader>as", "<Plug>(coc-codeaction-source)", opts)
    -- Remap keys for applying codeActions to the current buffer
    keyset("n", "<leader>ac", "<Plug>(coc-codeaction)", opts)
    -- Apply the most preferred quickfix action on the current line.
    keyset("n", "<leader>qf", "<Plug>(coc-fix-current)", opts)

    -- Remap keys for apply refactor code actions.
    keyset("n", "<leader>re", "<Plug>(coc-codeaction-refactor)", { silent = true })
    keyset("x", "<leader>r", "<Plug>(coc-codeaction-refactor-selected)", { silent = true })
    keyset("n", "<leader>r", "<Plug>(coc-codeaction-refactor-selected)", { silent = true })

    -- Run the Code Lens actions on the current line
    keyset("n", "<leader>ca", "<Plug>(coc-codelens-action)", opts)

    -- Map function and class text objects
    -- NOTE: Requires 'textDocument.documentSymbol' support from the language server
    keyset("x", "if", "<Plug>(coc-funcobj-i)", opts)
    keyset("o", "if", "<Plug>(coc-funcobj-i)", opts)
    keyset("x", "af", "<Plug>(coc-funcobj-a)", opts)
    keyset("o", "af", "<Plug>(coc-funcobj-a)", opts)
    keyset("x", "ic", "<Plug>(coc-classobj-i)", opts)
    keyset("o", "ic", "<Plug>(coc-classobj-i)", opts)
    keyset("x", "ac", "<Plug>(coc-classobj-a)", opts)
    keyset("o", "ac", "<Plug>(coc-classobj-a)", opts)

    -- Remap <C-f> and <C-b> to scroll float windows/popups
    ---@diagnostic disable-next-line: redefined-local
    local opts = { silent = true, nowait = true, expr = true }
    keyset("n", "<C-f>", 'coc#float#has_scroll() ? coc#float#scroll(1) : "<C-f>"', opts)
    keyset("n", "<C-b>", 'coc#float#has_scroll() ? coc#float#scroll(0) : "<C-b>"', opts)
    keyset("i", "<C-f>", 'coc#float#has_scroll() ? "<c-r>=coc#float#scroll(1)<cr>" : "<Right>"', opts)
    keyset("i", "<C-b>", 'coc#float#has_scroll() ? "<c-r>=coc#float#scroll(0)<cr>" : "<Left>"', opts)
    keyset("v", "<C-f>", 'coc#float#has_scroll() ? coc#float#scroll(1) : "<C-f>"', opts)
    keyset("v", "<C-b>", 'coc#float#has_scroll() ? coc#float#scroll(0) : "<C-b>"', opts)

    -- Use CTRL-S for selections ranges
    -- Requires 'textDocument/selectionRange' support of language server
    keyset("n", "<C-s>", "<Plug>(coc-range-select)", { silent = true })
    keyset("x", "<C-s>", "<Plug>(coc-range-select)", { silent = true })

    -- Add `:Format` command to format current buffer
    vim.api.nvim_create_user_command("Format", "call CocAction('format')", {})

    -- " Add `:Fold` command to fold current buffer
    vim.api.nvim_create_user_command("Fold", "call CocAction('fold', <f-args>)", { nargs = "?" })

    -- Add `:OR` command for organize imports of the current buffer
    vim.api.nvim_create_user_command("OR", "call CocActionAsync('runCommand', 'editor.action.organizeImport')", {})

    -- Add (Neo)Vim's native statusline support
    -- NOTE: Please see `:h coc-status` for integrations with external plugins that
    -- provide custom statusline: lightline.vim, vim-airline
    -- vim.opt.statusline:prepend("%{coc#status()}%{get(b:,'coc_current_function','')}")

    -- Mappings for CoCList
    -- code actions and coc stuff
    ---@diagnostic disable-next-line: redefined-local
    local opts = { silent = true, nowait = true }
    -- Show all diagnostics
    keyset("n", "<space>Ca", "<cmd>CocList diagnostics<cr>", opts)
    -- Manage extensions
    keyset("n", "<space>Ce", "<cmd>CocList extensions<cr>", opts)
    -- Show commands
    keyset("n", "<space>Cc", "<cmd>CocList commands<cr>", opts)
    -- Find symbol of current document
    keyset("n", "<space>Co", "<cmd>CocList outline<cr>", opts)
    -- Search workspace symbols
    keyset("n", "<space>Cs", "<cmd>CocList -I symbols<cr>", opts)
    -- Do default action for next item
    keyset("n", "<space>Cj", "<cmd>CocNext<cr>", opts)
    -- Do default action for previous item
    keyset("n", "<space>Ck", "<cmd>CocPrev<cr>", opts)
    -- Resume latest coc list
    keyset("n", "<space>Cp", "<cmd>CocListResume<cr>", opts)
  end,
}

File: .config/nvim/denops/config/plugins/coc.ts
// =============================================================================
// File        : coc.ts
// Author      : yukimemi
// Last Change : 2024/10/02 00:22:27.
// =============================================================================

import type { Plug } from "jsr:@yukimemi/dvpm@5.2.2";

import * as autocmd from "jsr:@denops/std@7.4.0/autocmd";
import * as fn from "jsr:@denops/std@7.4.0/function";
import * as lambda from "jsr:@denops/std@7.4.0/lambda";
import * as mapping from "jsr:@denops/std@7.4.0/mapping";
import * as op from "jsr:@denops/std@7.4.0/option";
import * as vars from "jsr:@denops/std@7.4.0/variable";
import { z } from "npm:zod@3.23.8";
import { execute } from "jsr:@denops/std@7.4.0/helper";

export const coc: Plug[] = [
  { url: "https://github.com/weirongxu/coc-explorer" },
  {
    url: "https://github.com/gelguy/wilder.nvim",
    afterFile: "~/.config/nvim/rc/after/wilder.lua",
  },
  {
    url: "https://github.com/neoclide/coc.nvim",
    dependencies: [
      "https://github.com/weirongxu/coc-explorer",
      "https://github.com/gelguy/wilder.nvim",
    ],
    rev: "release",
    before: async ({ denops }) => {
      await vars.g.set(denops, "coc_global_extensions", [
        "coc-deno",
        "coc-diagnostic",
        "coc-explorer",
        "coc-highlight",
        "coc-json",
        "coc-lists",
        "coc-lua",
        "coc-marketplace",
        "coc-nav",
        "coc-powershell",
        "coc-prettier",
        "coc-rust-analyzer",
        "coc-stylua",
        "coc-tsdetect",
        "coc-tsserver",
        "coc-snippets",
        "coc-vimlsp",
        "coc-xml",
      ]);
    },
    after: async ({ denops }) => {
      await mapping.map(
        denops,
        "<cr>",
        `coc#pum#visible() ? coc#pum#confirm() : '<c-g>u<cr><c-r>=coc#on_enter()<cr>'`,
        { mode: "i", noremap: true, expr: true, silent: true },
      );

      // Use <c-j> to trigger snippets
      await mapping.map(denops, "<c-j>", "<Plug>(coc-snippets-expand-jump)", {
        mode: "i",
      });
      await mapping.map(denops, "<c-j>", "<Plug>(coc-snippets-select)", {
        mode: "v",
      });
      // Use <leader>x for convert visual selected code to snippet
      await mapping.map(denops, "<leader>x", "<Plug>(coc-convert-snippet)", {
        mode: "x",
      });
      await vars.g.set(denops, "coc_snippet_next", "<c-j>");
      await vars.g.set(denops, "coc_snippet_prev", "<c-k>");

      await execute(
        denops,
        `
          inoremap <silent><expr> <TAB>
                \\ coc#pum#visible() ? coc#_select_confirm() :
                \\ coc#expandableOrJumpable() ? "\\<C-r>=coc#rpc#request('doKeymap', ['snippets-expand-jump',''])\\<CR>" :
                \\ CheckBackspace() ? "\\<TAB>" :
                \\ coc#refresh()

          function! CheckBackspace() abort
            let col = col('.') - 1
            return !col || getline('.')[col - 1]  =~# '\s'
          endfunction
        `,
      );

      await mapping.map(denops, "<c-n>", `coc#pum#visible() ? coc#pum#next(0) : "\\<c-n>"`, {
        expr: true,
        mode: "i",
        noremap: true,
        silent: true,
      });
      await mapping.map(denops, "<c-p>", `coc#pum#visible() ? coc#pum#prev(0) : "\\<c-p>"`, {
        expr: true,
        mode: "i",
        noremap: true,
        silent: true,
      });

      // Use <c-space> to trigger completion
      await mapping.map(denops, "<c-space>", "coc#refresh()", {
        expr: true,
        mode: "i",
        noremap: true,
        silent: true,
      });
      await mapping.map(denops, "[d", "<Plug>(coc-diagnostic-prev)", {
        silent: true,
        mode: "n",
      });
      await mapping.map(denops, "]d", "<Plug>(coc-diagnostic-next)", {
        silent: true,
        mode: "n",
      });

      // GoTo code navigation
      await mapping.map(denops, "gd", "<Plug>(coc-definition)", {
        silent: true,
        mode: "n",
      });
      await mapping.map(denops, "gy", "<Plug>(coc-type-definition)", {
        silent: true,
        mode: "n",
      });
      await mapping.map(denops, "gi", "<Plug>(coc-implementation)", {
        silent: true,
        mode: "n",
      });
      await mapping.map(denops, "gr", "<Plug>(coc-references)", {
        silent: true,
        mode: "n",
      });

      await mapping.map(
        denops,
        "K",
        `<cmd>call <SID>${denops.name}_notify("${
          lambda.register(
            denops,
            async () => {
              const cw = await fn.expand(denops, "<cword>");
              const ft = await op.filetype.getLocal(denops);
              if (["vim", "help"].some((t) => t === ft)) {
                await denops.cmd(`silent! h ${cw}`);
              } else if (
                z.number().parse(await denops.call(`coc#rpc#ready`)) === 1
              ) {
                await denops.call("CocActionAsync", "doHover");
              } else {
                await denops.cmd(`!${await op.keywordprg.get(denops)} ${cw}`);
              }
            },
          )
        }", [])<cr>`,
        { mode: "n", silent: true },
      );

      await autocmd.group(denops, "CocGroup", (helper) => {
        helper.remove("*");
        // Highlight the symbol and its references on a CursorHold event(cursor is idle)
        helper.define(
          "CursorHold",
          "*",
          `silent call CocActionAsync('highlight')`,
        );
        // Setup formatexpr specified filetype(s)
        helper.define(
          "FileType",
          "typescript,json",
          `setl formatexpr=CocAction('formatSelected')`,
        );
        // Update signature help on jump placeholder
        helper.define(
          "User",
          "CocJumpPlaceholder",
          `call CocActionAsync('showSignatureHelp')`,
        );
      });

      // Symbol renaming
      await mapping.map(denops, "<leader>rn", "<Plug>(coc-rename)", {
        mode: "n",
        silent: true,
      });

      // Formatting selected code
      await mapping.map(denops, "<leader>f", "<Plug>(coc-format-selected)", {
        mode: ["n", "x"],
        silent: true,
      });

      // Apply codeAction to the selected region
      // Example: `<leader>aap` for current paragraph
      await mapping.map(
        denops,
        "<leader>a",
        "<Plug>(coc-codeaction-selected)",
        { mode: ["n", "x"], silent: true },
      );

      // Remap keys for apply code actions at the cursor position.
      await mapping.map(denops, "<leader>ca", "<Plug>(coc-codeaction-cursor)", {
        mode: "n",
        silent: true,
      });
      // Remap keys for apply code actions affect whole buffer.
      await mapping.map(denops, "<leader>as", "<Plug>(coc-codeaction-source)", {
        mode: "n",
        silent: true,
      });
      // Remap keys for applying codeActions to the current buffer
      await mapping.map(denops, "<leader>ac", "<Plug>(coc-codeaction)", {
        mode: "n",
        silent: true,
      });
      // Apply the most preferred quickfix action on the current line.
      await mapping.map(denops, "<leader>qf", "<Plug>(coc-fix-current)", {
        mode: "n",
        silent: true,
      });

      // Remap keys for apply refactor code actions.
      await mapping.map(
        denops,
        "<leader>re",
        "<Plug>(coc-codeaction-refactor)",
        { mode: "n", silent: true },
      );
      await mapping.map(
        denops,
        "<leader>r",
        "<Plug>(coc-codeaction-refactor-selected)",
        { mode: "x", silent: true },
      );

      // Run the Code Lens actions on the current line
      await mapping.map(denops, "<leader>cl", "<Plug>(coc-codelens-action)", {
        mode: "n",
        silent: true,
      });

      // Map function and class text objects
      // NOTE: Requires 'textDocument.documentSymbol' support from the language server
      await mapping.map(denops, "if", "<Plug>(coc-funcobj-i)", {
        mode: ["x", "o"],
        silent: true,
      });

      await mapping.map(denops, "af", "<Plug>(coc-funcobj-a)", {
        mode: ["x", "o"],
        silent: true,
      });

      await mapping.map(denops, "ic", "<Plug>(coc-classobj-i)", {
        mode: ["x", "o"],
        silent: true,
      });

      await mapping.map(denops, "ac", "<Plug>(coc-classobj-a)", {
        mode: ["x", "o"],
        silent: true,
      });

      // Remap <C-f> and <C-b> to scroll float windows/popups
      await mapping.map(
        denops,
        "<C-f>",
        'coc#float#has_scroll() ? coc#float#scroll(1) : "<C-f>"',
        { mode: ["n", "v"], silent: true, nowait: true, expr: true },
      );
      await mapping.map(
        denops,
        "<C-b>",
        'coc#float#has_scroll() ? coc#float#scroll(0) : "<C-b>"',
        { mode: ["n", "v"], silent: true, nowait: true, expr: true },
      );
      await mapping.map(
        denops,
        "<C-f>",
        'coc#float#has_scroll() ? "<c-r>=coc#float#scroll(1)<cr>" : "<Right>"',
        { mode: "i", silent: true, nowait: true, expr: true },
      );
      await mapping.map(
        denops,
        "<C-b>",
        'coc#float#has_scroll() ? "<c-r>=coc#float#scroll(0)<cr>" : "<Left>"',
        { mode: "i", silent: true, nowait: true, expr: true },
      );

      // Use CTRL-S for selections ranges
      // Requires 'textDocument/selectionRange' support of language server
      await mapping.map(denops, "<C-s>", "<Plug>(coc-range-select)", {
        mode: ["n", "x"],
        silent: true,
      });

      // coc-lists.
      await mapping.map(denops, "<leader>cc", "<cmd>CocList<cr>", {
        mode: "n",
      });
      await mapping.map(
        denops,
        "<leader>co",
        "<cmd>CocList --auto-preview mru -A<cr>",
        {
          mode: "n",
        },
      );
      await mapping.map(
        denops,
        "<leader>cu",
        "<cmd>CocList --auto-preview mru<cr>",
        {
          mode: "n",
        },
      );
      await mapping.map(
        denops,
        "<leader>cb",
        "<cmd>CocList --auto-preview buffers<cr>",
        {
          mode: "n",
        },
      );

      // coc-explorer.
      await mapping.map(denops, "<space>e", "<cmd>CocCommand explorer<cr>", {
        mode: "n",
      });
    },
  },
];
